File: src/main/scala/com/resonant/core/graph/api/NodeElectric.java
package com.resonant.core.graph.api;

import com.resonant.core.graph.internal.Node;

/**
 * An electric node
 * @author Calclavia
 */
public interface NodeElectric extends Node<NodeElectric> {

	/**
	 * @return The resistance of the electric component in ohms.
	 */
	double resistance();

	/**
	 * @return The voltage (potential difference) of the component in volts.
	 */
	double voltage();

	/**
	 * @return The current of the component in amperes.
	 */
	double current();

	/**
	 * @return The power dissipated in the component.
	 */
	default double power() {
		return current() * voltage();
	}
}

File: src/main/scala/com/resonant/core/graph/internal/electric/NodeElectric.scala
package com.resonant.core.graph.internal.electric

import com.resonant.core.graph.api.NodeProvider
import com.resonant.core.graph.internal.NodeBlockConnect
import com.resonant.wrapper.core.api.tile.DebugInfo

import scala.beans.BeanProperty
import scala.collection.convert.wrapAll._

/**
 * @author Calclavia
 */
abstract class NodeElectric(parent: NodeProvider) extends NodeBlockConnect[NodeElectric](parent) with DebugInfo with com.resonant.core.graph.api.NodeElectric {
	@BeanProperty
	var resistance = 1d

	override def getDebugInfo = List(toString)
}
